vizro-core/src/vizro/models/_utils.py

```diff
@@ -0,0 +1,167 @@
+
+    if isinstance(object, dict) and "__vizro_model__" in object:
+        __vizro_model__ = object.pop("__vizro_model__")
+
+        # This is required to back-engineer the actions chains. It is easier to handle in the string conversion here
+        # than in the dict creation, because we end up with nested lists when being forced to return a list of actions.
+        # If we handle it in dict of vm.BaseModel, then we created an unexpected dict return.
+        if __vizro_model__ == ACTIONS_CHAIN:
+            action_data = object[ACTION]
+            return ", ".join(_dict_to_python(item) for item in action_data)
+        else:
+            # This is very similar to doing repr but includes the vm. prefix and calls _object_to_python_code
+            # rather than repr recursively.
+            fields = ", ".join(f"{field_name}={_dict_to_python(value)}" for field_name, value in object.items())
+            return f"vm.{__vizro_model__}({fields})"
+    elif isinstance(object, dict):
+        fields = ", ".join(f"{field_name}={_dict_to_python(value)}" for field_name, value in object.items())
+        return "{" + fields + "}"
+    elif isinstance(object, list):
+        # Need to do this manually to avoid extra quotation marks that arise when doing repr(List).
+        code_string = ", ".join(_dict_to_python(item) for item in object)
+        return f"[{code_string}]"
+    elif isinstance(object, CapturedCallable):
+        return object._repr_clean()
+    else:
+        return repr(object)
+
+
+def _concatenate_code(
+    code: str,
+    extra_imports: Optional[str] = None,
+    callable_defs: Optional[str] = None,
+    data_settings: Optional[str] = None,
+) -> str:
+    callable_defs = CALLABLE_TEMPLATE.format(callable_defs=callable_defs) if callable_defs else ""
+    data_settings = DATA_TEMPLATE.format(data_setting=data_settings) if data_settings else ""
+    unformatted_code = TO_PYTHON_TEMPLATE.format(
+        code=code,
+        extra_imports=extra_imports if extra_imports else "",
+        callable_defs=callable_defs,
+        data_settings=data_settings,
+    )
+    return _format_and_lint(unformatted_code)
+
+# The two extract helper functions may not work when we refactor the model_manager to work differently when models
+# are created. An alternative approach to iterating through the model_manager is to recurse through the object as
+# is done in the _dict_to_python function.
+# Note also that these functions find also unintended model_manager additions, a known but accepted limitation.
+def _extract_captured_callable_source() -> Set[str]:
+    from vizro.models.types import CapturedCallable
+
+    captured_callable_sources = set()
+    for model_id in model_manager:
+        for _, value in model_manager[model_id]:
+            if isinstance(value, CapturedCallable) and all(
+                replacement.original not in value._function.__module__ for replacement in REPLACEMENT_STRINGS
+            ):
```

Comment:
Just to understand, why is the comparison here substring rather than `==`?

i.e. if we changed `REPLACEMENT_STRINGS` then why would this be wrong?
```suggestion
            if isinstance(value, CapturedCallable) and value._function.__module__ not in REPLACEMENT_STRINGS):
```

Reply:
Similar reason as to above, I would like to avoid trying to match exact module paths, like `plotly.express._graph_obj` or similar. It bears a small risk, but will overall be more robust if there is alternative to import paths, but when the main content machtes

Reply:
Oh ok, I see 👍 Thanks.